File: app/src/main/java/com/moggot/findmycarlocation/presentation/map/MapPresenter.java
package com.moggot.findmycarlocation.presentation.map;

import android.support.annotation.NonNull;

import com.google.android.gms.maps.model.LatLng;
import com.google.maps.android.PolyUtil;
import com.moggot.findmycarlocation.data.model.parking.ParkingModel;
import com.moggot.findmycarlocation.domain.LocationInteractor;
import com.moggot.findmycarlocation.domain.MainInteractor;
import com.moggot.findmycarlocation.domain.MapInteractor;
import com.moggot.findmycarlocation.presentation.common.BasePresenter;

import java.util.List;

import javax.inject.Inject;

public class MapPresenter extends BasePresenter<MapView> {

    @NonNull
    private final MapInteractor mapInteractor;
    @NonNull
    private final LocationInteractor locationInteractor;
    @NonNull
    private final MainInteractor mainInteractor;

    @Inject
    public MapPresenter(@NonNull MapInteractor mapInteractor,
                        @NonNull LocationInteractor locationInteractor,
                        @NonNull MainInteractor mainInteractor) {
        this.mapInteractor = mapInteractor;
        this.locationInteractor = locationInteractor;
        this.mainInteractor = mainInteractor;
    }

    public void buildRoute() {
        if (getView() == null) {
            return;
        }
        unSubscribeOnDetach(locationInteractor.getLocation()
                .flatMap(location -> {
                    LatLng originLocation = new LatLng(location.getLatitude(), location.getLongitude());
                    return mapInteractor.getRoute(originLocation)
                            .filter(path -> !path.getRoutes().isEmpty())
                            .doOnSuccess(route -> {
                                getView().showAd();
                                String pointsStr = route.getRoutes().get(0).getOverviewPolyline().getPoints();
                                List<LatLng> points = PolyUtil.decode(pointsStr);
                                getView().showRoute(points);
                                String distance = route.getRoutes().get(0).getLegs().get(0).getDistance().getText();
                                String duration = route.getRoutes().get(0).getLegs().get(0).getDuration().getText();
                                getView().showDistance(distance);
                                getView().showDuration(duration);
                            });
                }).doOnError(throwable -> getView().showError())
                .subscribe());
    }

    public void drawCircle() {
        if (getView() != null) {
            ParkingModel parkingModel = mainInteractor.loadParkingData();
            getView().decoratePoint(parkingModel.getLocation());
        }
    }

    public void foundCar() {
        mainInteractor.markCarIsFound();
    }
}

File: app/src/main/java/com/moggot/findmycarlocation/map/MapPresenter.kt
package com.moggot.findmycarlocation.map

import com.google.android.gms.maps.model.LatLng
import com.moggot.findmycarlocation.base.BasePresenter
import com.moggot.findmycarlocation.home.LocationInteractor
import com.moggot.findmycarlocation.home.MainInteractor
import com.moggot.findmycarlocation.retry.RetryManager
import timber.log.Timber

import javax.inject.Inject

class MapPresenter @Inject
constructor(private val mapInteractor: MapInteractor,
            private val locationInteractor: LocationInteractor,
            private val mainInteractor: MainInteractor,
            private val retryManager: RetryManager) : BasePresenter<GoogleMapView>() {

    fun buildRoute() {
        if (!mainInteractor.loadParkingData().isParking) {
            return
        }
        Timber.d("buildRoute")
        disposable.add(locationInteractor.location
                .flatMap { location ->
                    val originLocation = LatLng(location.latitude, location.longitude)
                    mapInteractor.getRoute(originLocation)
                            .filter { path -> path.routes.isNotEmpty() }
                            .doOnSuccess {
                                Timber.d("doOnSuccess")
                                view { drawRoute(it) } }
                }.doOnError {
                    Timber.d("onerror")
                    view { onError(it) }
                }
                .toObservable()
                .retryWhen { retryHandler -> retryHandler.flatMap(retryManager::observeRetries) }
                .subscribe())
    }

    fun getCurrentLocation() {
        disposable.add(locationInteractor.location
                .doOnError { throwable -> view { onError(throwable) } }
                .subscribe { view { updateLocation(it) } })
    }

    fun drawCircle(): LatLng {
        return mainInteractor.loadParkingData().location
    }

    fun retryCall() {
        retryManager.retry()
    }

    fun foundCar() {
        mainInteractor.markCarIsFound()
        view { onCarFound() }
    }
}
